### PR TITLE
Add explicit centers and radius to gradients

### PR DESCRIPTION
**Pull request by tilladam**
## Summary
- add explicit `at <x> <y>` centers for radial and conic gradients
- add explicit `size <radius>` support for radial gradients
- wire center/radius metadata through compiler, generated code, interpreter, C++ API, and renderers
- document the new syntax and add focused syntax, driver, and screenshot coverage

## Tests
- `cargo test -p i-slint-compiler --features display-diagnostics --test syntax_tests`
- `SLINT_TEST_FILTER=gradient_at_center cargo test -p test-driver-interpreter`
- `SLINT_TEST_FILTER=gradient_at_center cargo test -p test-driver-cpp`

### CODE REVIEW

**Comment by ogoffart** [comment]
Thanks. I had a quick look at the proposed syntax.
The `size <radius>` syntax doesn't match CSS. In CSS you'd write `circle 80px at 50px 50px` (the radius comes right after circle, before at, with no extra keyword). 
We try to follow the CSS spec so that one can copy paste from other tool that generate CSS syntac. So let's follow the spec unless there is a good reason not to.

**Comment by tilladam** [comment]
> We try to follow the CSS spec so that one can copy paste from other tool that generate CSS syntac. So let's follow the spec unless there is a good reason not to.

Ok, I've adjusted to the CSS syntax now.

File: api/cpp/include/private/slint_brush.h [hunk 1]
```diff
@@ -209,7 +219,10 @@ class Brush
     /// color that is identical to \a a's color. If it holds a gradient, then the gradients must be
     /// identical. Returns false if the brushes differ in what they hold or their respective color
     /// or gradient are not equal.
-    friend bool operator==(const Brush &a, const Brush &b) { return a.data == b.data; }
+    friend bool operator==(const Brush &a, const Brush &b)
```
Comment (ogoffart) [inline api/cpp/include/private/slint_brush.h:254]: I'm curious: what was wrong with the auto-generated `==`?

File: api/cpp/include/private/slint_brush.h [hunk 2]
```diff
@@ -209,7 +219,10 @@ class Brush
...
+    {
+        return cbindgen_private::types::slint_brush_compare_equal(&a.data, &b.data);
+    }
     /// Returns false if \a is not equal to \a b; true otherwise.
     friend bool operator!=(const Brush &a, const Brush &b) { return a.data != b.data; }
```
Comment (ogoffart) [inline api/cpp/include/private/slint_brush.h]: If the `==` was wrong, the `!=` is probably also wrong

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,14 @@ All notable changes to this project are documented in this file.
 
 ## [Unreleased]
 
+### Slint
+
+ - `@radial-gradient` and `@conic-gradient` now support explicit `at <x> <y>` centers in local coordinates. `@radial-gradient` also supports `circle <radius>`.
+
+### Breaking changes
+
+ - **ABI break**: `RadialGradientBrush` and `ConicGradientBrush` gained new `center_x`, `center_y`, and (for radial) `radius` fields. Binaries compiled against a previous version of `i-slint-core` must be recompiled.
```
Comment (ogoffart) [inline CHANGELOG.md]: This is internal API anyway.

Please put the changelog in the commit message instead.
individual commits don't need to edit the changelog.

File: internal/core/graphics/brush.rs [hunk 4]
```diff
@@ -244,28 +253,161 @@ impl LinearGradientBrush {
...
+    stops: SharedVector<GradientStop>,
+    /// Explicit horizontal center in the element's local coordinate space.
+    /// `f32::NAN` means "use the element's bounding box center (width / 2)".
+    pub center_x: f32,
+    /// Explicit vertical center in the element's local coordinate space.
+    /// `f32::NAN` means "use the element's bounding box center (height / 2)".
+    pub center_y: f32,
+    /// Explicit radius in the element's local coordinate space.
+    /// A negative value means "use half of the element's bounding box diagonal".
+    pub radius: f32,
```
Comment (ogoffart) [inline internal/core/graphics/brush.rs]: RadialGradientBrush is actually exposed to the public API via the `Brush` enum.  
But i'd rather we keep this kind of things internal.
(These fields and also the new assiciated functions)
Is `pub (crate)` not enough?
could that be done with helper free functions instead?

File: internal/core/graphics/brush.rs [hunk 5]
```diff
@@ -244,28 +253,161 @@ impl LinearGradientBrush {
...
+}
+
+/// The RadialGradientBrush describes a way of filling a shape with a circular gradient.
+///
+/// When `center_x` / `center_y` are `f32::NAN` the gradient is centered in the element's
+/// bounding box (`width / 2`, `height / 2`). When `radius` is negative the radius defaults
+/// to half of the element's bounding box diagonal.
+#[derive(Clone, Debug)]
+#[repr(C)]
+pub struct RadialGradientBrush {
```
Comment (ogoffart) [inline internal/core/graphics/brush.rs]: This change will more than double the size of `Brush`.
We've stored the angle of the linear and conical gradients as the first stop to save size.
Maybe we don't need to save size and this optimization was overkill, or we could do the same and store the center and radius in the stops.

File: tests/screenshots/references/software/basic/radial-gradients.png
Comment (ogoffart) [inline tests/screenshots/references/software/basic/radial-gradients.png:1]: The black dot in the center of the last gradient in the bottom right was perhaps a bug.
It is a "// Edge case: invisible stop at 0%"  and that stop wasn't invisible before apparently.

File: internal/renderers/software/scene.rs [hunk 6]
```diff
@@ -537,9 +537,12 @@ pub struct LinearGradientCommand {
 pub struct RadialGradientCommand {
     /// The gradient stops (colors and positions)
     pub stops: i_slint_core::SharedVector<i_slint_core::graphics::GradientStop>,
-    /// Center of the gradient relative to the item position
-    pub center_x: PhysicalLength,
-    pub center_y: PhysicalLength,
+    /// Center in physical pixels, relative to the clipped rect's top-left corner.
+    /// Stored as f32 to avoid i16 saturation for off-bbox centers at high scale factors.
```
Comment (ogoffart) [inline internal/renderers/software/scene.rs:541]: The goal is to avoid per-pixel floating point operations in the tight loop (in draw_functions.rs).
Should it be stored in `i32` or in `Fixed<i32, 8>`?

That said, there is already a few floating point computation in the radial gradient rendering, so perhaps this doesn't matter.

File: internal/renderers/software/scene.rs
Comment (ogoffart) [inline internal/renderers/software/scene.rs:555]: Any reason the conic gradient algorithm was changed?

File: internal/core/graphics/brush.rs [hunk 7]
```diff
@@ -244,28 +253,161 @@ impl LinearGradientBrush {
...
+/// Shared center resolution for the logical and scaled radial/conic methods.
+/// When `scale_factor` is 1.0 this is identical to the unscaled case.
+#[inline]
+fn center_or_bbox(cx: f32, cy: f32, width: f32, height: f32, scale_factor: f32) -> (f32, f32) {
+    if cx.is_nan() { (width / 2.0, height / 2.0) } else { (cx * scale_factor, cy * scale_factor) }
+}
+
+/// The RadialGradientBrush describes a way of filling a shape with a circular gradient.
+///
+/// When `center_x` / `center_y` are `f32::NAN` the gradient is centered in the element's
```
Comment (ogoffart) [inline internal/core/graphics/brush.rs]: I'm not particulary a fan of this f32::NaN encoding.

The CSS also supports the very useful `%` as unit for the center, so we could also support that, and we could encode that in a type like so

```
#[repr(C)]
enum RelativeOrAbsolute {
   Relative(f32),
   Absolute(LogicalLenght),
}
``` 

We could also try to encode the thing in a i32 using bit masks and stuff, but maybe not worth it.

Comment (tilladam) [reply internal/core/graphics/brush.rs]: Addressed in the latest push: center_x/center_y/radius are no longer struct fields at all. Following the same pattern as LinearGradientBrush (which stores the angle in stop[0].position), both RadialGradientBrush and ConicGradientBrush are now #[repr(transparent)] over SharedVector<GradientStop>, with the first 3 fake stops encoding the metadata:

RadialGradientBrush: stop[0].position=center_x, [1].position=center_y, [2].position=radius, [3+]=real stops
ConicGradientBrush:  stop[0].position=angle (existing), [1].position=center_x, [2].position=center_y, [3+]=real stops

NaN/negative sentinels are now an internal encoding detail only — no public fields, no cbindgen-generated broken comparisons. The RelativeOrAbsolute approach would be cleaner if we later want CSS-style % units, but that's a bigger change; this PR focuses on px-based explicit center/radius.

File: tests/screenshots/references/software/basic/radial-gradients.png
Comment (tilladam) [reply tests/screenshots/references/software/basic/radial-gradients.png:1]: Investigated this. The black dot in the old reference was a side-effect of the original implementation using integer truncation for the center: (absolute_center_x - clipped.min_x()) as i16. For a cell whose width is an odd number of pixels, width/2 truncates, placing the center exactly on a pixel boundary. That pixel landed at position=0.0 → transparent → appeared black on the black background.

The PR uses float centers throughout, so no pixel happens to sit at exactly position=0.0. The center area instead renders the correct interpolated color (near-transparent dark red for this gradient). The reference PNG update is correct.

File: internal/renderers/software/scene.rs
Comment (tilladam) [reply internal/renderers/software/scene.rs:555]: The core algorithm (atan2, angle normalization, stop interpolation) is unchanged. The only difference is that the center is now read from g.center_x/g.center_y in the command rather than being computed inline as rect.width() / 2. That change is necessary because the center can now be explicit (set by the user), so the draw function can no longer assume it's always the bbox midpoint.

File: internal/core/graphics/brush.rs (hunk 5 above)
Comment (tilladam) [reply internal/core/graphics/brush.rs]: Done — both types are now #[repr(transparent)] over SharedVector<GradientStop>, same as LinearGradientBrush. No size increase to Brush.

File: internal/core/graphics/brush.rs (hunk 4 above)
Comment (tilladam) [reply internal/core/graphics/brush.rs]: Done — center_x, center_y, and radius are no longer fields at all (they're encoded in the leading fake stops). Public surface is now only the builder methods with_center() and with_radius(), and the resolution methods center_or_default() / radius_or_default().

File: api/cpp/include/private/slint_brush.h (hunk 1 above)
Comment (tilladam) [reply api/cpp/include/private/slint_brush.h:254]: Good question. The cbindgen-generated `==` does a plain field-wise comparison, which means IEEE float equality on the new `center_x`/`center_y`/`radius` fields. Those fields use NaN as a "use bounding-box default" sentinel, and NaN != NaN under IEEE 754, so two brushes both using the default center would have compared unequal. The custom `slint_brush_compare_equal` treats sentinel values as equal. Added a `\note` on `operator==` in 6910555a5 explaining this.

File: api/cpp/include/private/slint_brush.h (hunk 2 above)
Comment (tilladam) [reply api/cpp/include/private/slint_brush.h]: Right — `!=` was delegating to the cbindgen-generated `!=`, which had the same NaN problem. Fixed in 6910555a5 to delegate to `!(a == b)` so it goes through the NaN-aware path.

File: CHANGELOG.md (hunk 3 above)
Comment (tilladam) [reply CHANGELOG.md]: Done in 6910555a5 — dropped the changelog hunk and put the ABI-break note in the commit message instead.

File: internal/renderers/software/scene.rs (hunk 6 above)
Comment (tilladam) [reply internal/renderers/software/scene.rs:541]: Note that e6748a937 later changed the encoding entirely (center is now read from the leading fake stops), but the per-pixel question still applies to `RadialGradientCommand`. The field is now `PhysicalLength`, which the software renderer defines as `euclid::Length<i16, PhysicalPx>` (`renderers/software/lib.rs:59`) — same integer type used elsewhere in the renderer, so no per-pixel floats in this path.

File: internal/compiler/llr/expression.rs [hunk 8]
```diff
@@ -173,13 +173,22 @@ pub enum Expression {
     },
 
     RadialGradient {
+        /// Explicit gradient center in the element's local coordinate space (`at <x> <y>`).
+        /// `None` means use the element's bbox centre.
+        center: Option<(Box<Expression>, Box<Expression>)>,
+        /// Explicit radius in the element's local coordinate space (`size <r>`).
```
Comment (ogoffart) [inline internal/compiler/llr/expression.rs]:
```suggestion
        /// Explicit radius in the element's local coordinate space (`circle <radius>`).
```

Is that right?

File: internal/renderers/skia/itemrenderer.rs [hunk 9]
```diff
@@ -771,6 +780,7 @@ impl ItemRenderer for SkiaItemRenderer<'_> {
                 path.fill(),
                 PhysicalLength::new(viewbox_width),
                 PhysicalLength::new(viewbox_height),
+                1.0,
```
Comment (ogoffart) [inline internal/renderers/skia/itemrenderer.rs:783]: Is 1.0 correct here?

Comment (tilladam) [reply internal/renderers/skia/itemrenderer.rs:783]: Yes it is, added a comment.
